contributors/devel/sig-instrumentation/migration-to-structured-logging.md

```diff
@@ -0,0 +1,329 @@
+
+With `klog` structured logging borrowing interface from [logr] it also inherits it's differences in semantic of error
+function. This is visible in `klog.ErrorS` which requires passing an `error` as first argument. This is because in
+[logr] Error is used to inform users about errors received from subordinate function calls.
+This means that `klog.ErrorS` should only be used when there is a obvious Golang `error` object available.
+
+Creating error just to call `klog.ErrorS` or passing `nil` is discouraged, please consider using `klog.InfoS` instead.
+
+### Replacing Fatal calls
+
+Usage of Fatal should be discouraged and it's not available in new functions . Instead of depending on logger to exist
+process you should consider returning error and ensuring it is properly handled.
+
+## Remove string formatting from log message
+
+With structured logging, log messages are no longer formatted, leaving argument marshalling up to logging client
+implementation. This allows messages to be a static description of event.
+
+All string formatting (`%d`, `%v`, `%w`, `%s`) should be removed and message simplified.
+Describing arguments in log messages is no longer needed and should be removed leaving only description of what happen.
+
+For example
+```go
+klog.Infof("delete object %s with propagation policy %s", ...)
+```
+should be changed to
+```go
+klog.Infof("delete object", ...)
+```
+
+Some logs are constructed solely from string formats. In those cases a message needs to be derived from context of
+log call.
+
+For example http access logs
+```go
+func LogHTTP(r *http.Request) {
+   klog.Infof("%s %s: (%v) %v%v%v [%s %s]", ...)
+}
+```
+should be changed to
+```go
+func LogHTTP(r *http.Request) {
+   klog.InfoS("HTTP", ...)
+}
+```
+
+### Name arguments
+
+Even thou new structured logging functions have very similar function prototype `func (string, ...interface{}) it's
+has different meaning for variadic arguments. Instead just passing arguments, now we are passing key value pairs of
+argument name and argument value. This means when migrating log call we need to add additional string before each
+argument, that will be used as it's name.
+
+For example
+```go
+func LogHTTP(r *http.Request) {
+   klog.Infof("Got request, path: %s, method: %s", r.Path, r.Method)
+}
+```
+should be changed to
+```go
+func LogHTTP(r *http.Request) {
+   klog.InfoS("Got request", "path", r.Path, "method", r.Method)
+}
+```
+
+Names of arguments should use [lowerCamelCase] and be alphanumeric. Arguments names in one log call should be unique.
+Names should be picked based on semantic meaning of value itself, not the context in which is used (log message should
+imply the context). For example names like `status` should be used over (`desiredStatus`, `oldStatus`, `badStatus`) thus
+allowing to query and join different logs lines on `status` field.
+
+Kubernetes objects should be referenced using only their kind, no matter their api group or version. Example argument
```

Comment:
What about the situation when the kind of an object is not known up front (for instance metav1.Object is passed as argument to function)? Should it be logged as "object" or rather the actual object kind should be used.

Reply:
Good question. I think best option would be to add new function to `klog` that would generate key based on `Kind`. But I would wait util we see some examples of this. For now I would recommend using "object"

Reply:
Agree.